File: box/src/main/java/de/qabel/box/storage/FolderNavigation.java
package de.qabel.box.storage;

import de.qabel.core.crypto.QblECKeyPair;
import de.qabel.box.storage.exceptions.QblStorageException;
import de.qabel.box.storage.exceptions.QblStorageNotFound;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.spongycastle.crypto.params.KeyParameter;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.security.InvalidKeyException;
import java.util.Arrays;
import java.util.Map;
import java.util.WeakHashMap;

public class FolderNavigation extends AbstractNavigation {
    private Map<Integer, String> directoryMetadataMHashes = new WeakHashMap<>();

    private static final Logger logger = LoggerFactory.getLogger(FolderNavigation.class);

    private final byte[] key;

    public FolderNavigation(String prefix, DirectoryMetadata dm, QblECKeyPair keyPair, byte[] key, byte[] deviceId,
                            StorageReadBackend readBackend, StorageWriteBackend writeBackend, IndexNavigation indexNavigation) {
        super(prefix, dm, keyPair, deviceId, readBackend, writeBackend, indexNavigation);
        this.key = key;
    }

    @Override
    protected void uploadDirectoryMetadata() throws QblStorageException {
        logger.trace("Uploading directory metadata");
        KeyParameter secretKey = new KeyParameter(key);
        uploadEncrypted(getDm().getPath(), secretKey, getDm().getFileName());
    }

    @Override
    public DirectoryMetadata reloadMetadata() throws QblStorageException {
        logger.trace("Reloading directory metadata");
        // duplicate of navigate()
        try (StorageDownload download = getReadBackend().download(getDm().getFileName(), getMHash())) {
            InputStream indexDl = download.getInputStream();
            File tmp = File.createTempFile("dir", "db7", getDm().getTempDir());
            tmp.deleteOnExit();
            KeyParameter key = new KeyParameter(this.key);
            if (getCryptoUtils().decryptFileAuthenticatedSymmetricAndValidateTag(indexDl, tmp, key)) {
                DirectoryMetadata newDM = DirectoryMetadata.openDatabase(tmp, getDeviceId(), getDm().getFileName(), getDm().getTempDir());
                directoryMetadataMHashes.put(Arrays.hashCode(newDM.getVersion()), download.getMHash());
                return newDM;
            } else {
                throw new QblStorageNotFound("Invalid key");
            }
        } catch (UnmodifiedException e) {
            return getDm();
        } catch (IOException | InvalidKeyException e) {
            throw new QblStorageException(e);
        }
    }

    private String getMHash() throws QblStorageException {
        return directoryMetadataMHashes.get(Arrays.hashCode(getDm().getVersion()));
    }
}

File: box/src/main/java/de/qabel/box/storage/FolderNavigation.kt
package de.qabel.box.storage

import de.qabel.box.storage.dto.BoxPath
import de.qabel.box.storage.exceptions.QblStorageException
import de.qabel.box.storage.exceptions.QblStorageNotFound
import org.slf4j.LoggerFactory
import org.spongycastle.crypto.params.KeyParameter
import java.io.File
import java.io.IOException
import java.security.InvalidKeyException
import java.util.*

class FolderNavigation(
        override val path: BoxPath.FolderLike,
        dm: DirectoryMetadata,
        val key: ByteArray,
        override val indexNavigation: IndexNavigation,
        volumeConfig: BoxVolumeConfig
) : AbstractNavigation(path, dm, volumeConfig) {
    private val directoryMetadataMHashes = WeakHashMap<Int, String>()
    private val logger by lazy { LoggerFactory.getLogger(FolderNavigation::class.java) }

    @Throws(QblStorageException::class)
    override fun uploadDirectoryMetadata() {
        logger.trace("Uploading directory metadata")
        uploadEncrypted(dm.path, KeyParameter(key), dm.fileName)
    }

    @Throws(QblStorageException::class)
    override fun reloadMetadata(): DirectoryMetadata {
        logger.trace("Reloading directory metadata $path")
        // duplicate of navigate()
        try {
            readBackend.download(dm.fileName, mHash).use { download ->
                val indexDl = download.inputStream
                val tmp = File.createTempFile("dir", "db7", tempDir)
                tmp.deleteOnExit()
                val key = KeyParameter(this.key)
                if (cryptoUtils.decryptFileAuthenticatedSymmetricAndValidateTag(indexDl, tmp, key)) {
                    val newDM = directoryFactory.open(tmp, dm.fileName)
                    directoryMetadataMHashes.put(Arrays.hashCode(newDM.version), download.mHash)
                    return newDM
                } else {
                    throw QblStorageNotFound("Invalid key")
                }
            }
        } catch (e: UnmodifiedException) {
            return dm
        } catch (e: IOException) {
            throw QblStorageException(e)
        } catch (e: InvalidKeyException) {
            throw QblStorageException(e)
        }
    }

    private val mHash: String?
        @Throws(QblStorageException::class)
        get() = directoryMetadataMHashes[Arrays.hashCode(dm.version)]
}
